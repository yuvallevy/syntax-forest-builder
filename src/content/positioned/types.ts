import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type Width = number;

export type PlotCoords = {
  plotX: PlotX;
  plotY: PlotY;
};

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: Id[];
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  slice: StringSlice;
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedStrandedNode = NodeCommon & WithPositionInTree;

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode | PositionedStrandedNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: PlotCoords;
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
