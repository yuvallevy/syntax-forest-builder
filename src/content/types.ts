export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;

export type StringSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type NodeIndicatorInPlot = {
  treeId: Id;
  nodeId: Id;
};

export type TreeCommon = {
  sentence: Sentence;
}

export type NodeCommon = {
  label: NodeLabel;
};
