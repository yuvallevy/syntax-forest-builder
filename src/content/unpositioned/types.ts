import { Id, IdMap, NodeCommon, StringSlice, TreeCommon } from '../types';

type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;

export type PlotCoordsOffset = {
  dPlotX: DPlotX;
  dPlotY: DPlotY;
};

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: PlotCoordsOffset;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: Id[];
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: StringSlice;
  triangle: boolean;
};

export type UnpositionedStrandedNode = NodeCommon & WithOffsetInTree & {
  formerDescendants?: IdMap<UnpositionedNode>;
  formerSlice?: StringSlice;
  formerlyTriangle?: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode | UnpositionedStrandedNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export const isTerminal = (node: UnpositionedNode): node is UnpositionedTerminalNode =>
  'slice' in node;
