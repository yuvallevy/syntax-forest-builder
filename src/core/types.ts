export type Id = string;
export type Sentence = string;
export type NodeLabel = string;

type SliceStart = number;
type SliceEndExclusive = number;
type PlotX = number;
type PlotY = number;
type TreeX = number;
type TreeY = number;
type DPlotX = number;
type DPlotY = number;
type DTreeX = number;
type DTreeY = number;
type Width = number;

export type NodeSlice = [SliceStart, SliceEndExclusive];

export type IdMap<T> = Record<Id, T>;

export type UnpositionedForest = { plots: IdMap<UnpositionedPlot>; };

export type UnpositionedPlot = { trees: IdMap<UnpositionedTree>; };

type TreeCommon = {
  sentence: Sentence;
}

export type UnpositionedTree = TreeCommon & {
  nodes: IdMap<UnpositionedNode>;
  offset: { dPlotX: DPlotX; dPlotY: DPlotY; };
};

export type NodeCommon = {
  label: NodeLabel;
};

type WithOffsetInTree = {
  offset: {
    dTreeX: DTreeX;
    dTreeY: DTreeY;
  };
};

export type UnpositionedBranchingNode = NodeCommon & WithOffsetInTree & {
  children: IdMap<UnpositionedNode>;
};

export type UnpositionedTerminalNode = NodeCommon & WithOffsetInTree & {
  slice: NodeSlice;
  triangle: boolean;
};

export type UnpositionedNode = UnpositionedBranchingNode | UnpositionedTerminalNode;

export const isBranching = (node: UnpositionedNode): node is UnpositionedBranchingNode =>
  'children' in node;

export type PositionInTree = {
  treeX: TreeX;
  treeY: TreeY;
};

type WithPositionInTree = { position: PositionInTree };

export type PositionedBranchingNode = NodeCommon & WithPositionInTree & {
  children: IdMap<PositionedNode>;
};

export type PositionedTerminalNode = NodeCommon & WithPositionInTree & {
  triangle?: {
    treeX1: TreeX;
    treeX2: TreeX;
  }
};

export type PositionedNode = PositionedBranchingNode | PositionedTerminalNode;

export type PositionedTree = TreeCommon & {
  nodes: IdMap<PositionedNode>;
  position: {
    plotX: PlotX;
    plotY: PlotY;
  };
  width: Width;
};

export type PositionedPlot = { trees: IdMap<PositionedTree> };
