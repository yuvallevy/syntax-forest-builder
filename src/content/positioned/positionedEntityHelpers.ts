import { associateWith, filterEntries, isEmpty, mapValues } from '../../util/objTransforms';
import { Id, IdMap, StringSlice } from '../types';
import slicesOverlap from '../slicesOverlap';
import { PositionedNode, PositionedTree } from './types';

/**
 * Returns nodes matching the given predicate in the given tree.
 */
export const filterPositionedNodesInTree =
  (predicate: (node: PositionedNode) => boolean) =>
  (tree: PositionedTree): IdMap<PositionedNode> =>
    filterEntries(tree.nodes, ([_, node]) => predicate(node));

/**
 * Returns nodes with the given IDs in the given tree.
 */
export const filterPositionedNodesInTreeById =
  (nodeIds: Id[]) =>
  (tree: PositionedTree): IdMap<PositionedNode> =>
    filterEntries(tree.nodes, ([nodeId, _]) => nodeIds.includes(nodeId));

/**
 * Returns whether the node with the given ID is a top-level node in the given node ID map.
 */
export const isTopLevel = (nodes: IdMap<PositionedNode>) => (nodeId: Id) =>
  isEmpty(filterEntries(nodes, ([_, node]) => 'children' in node && node.children.includes(nodeId)));

/**
 * Returns an ID map consisting of the top-level nodes in the given tree.
 */
export const getTopLevelPositionedNodes = (tree: PositionedTree): IdMap<PositionedNode> =>
  filterEntries(tree.nodes, ([nodeId, _]) => isTopLevel(tree.nodes)(nodeId));

/**
 * Receives an array of node IDs in the given tree and returns them sorted by X-coordinate.
 */
export const sortPositionedNodesByXCoord = (tree: PositionedTree) => (nodeIds: Id[]) =>
  Object.entries(associateWith(nodeIds, nodeId => tree.nodes[nodeId].position.treeX))
    .sort(([_, treeX1], [__, treeX2]) => treeX1 - treeX2)
    .map(([nodeId, _]) => nodeId);

/**
 * Returns true if there are no nodes assigned to any slices overlapping the given one.
 */
export const isSliceUnassigned = (tree: PositionedTree) => (slice: StringSlice) => {
  const nodesAssignedToSlice = mapValues(tree.nodes, node => 'slice' in node && slicesOverlap(node.slice, slice));
  return nodesAssignedToSlice.every(overlapsSlice => !overlapsSlice);
};
