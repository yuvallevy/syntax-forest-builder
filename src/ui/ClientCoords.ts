type ClientX = number;
type ClientY = number;

type ClientCoords = {
  clientX: ClientX;
  clientY: ClientY;
};

export type ClientRect = {
  topLeft: ClientCoords,
  bottomRight: ClientCoords,
};

export default ClientCoords;
