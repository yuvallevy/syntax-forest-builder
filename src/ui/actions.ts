import { union } from '../core/objTransforms';
import { Id } from '../core/types';
import { UiState } from './state';

type UiAction =
  | { type: 'selectNodes', plotId: Id, treeIds: Id[], nodeIds: Id[], mode: 'set' | 'add' }
;

export const reducer = (state: UiState, action: UiAction): UiState => {
  switch (action.type) {
    case 'selectNodes':
      return {
        ...state,
        activePlotId: action.plotId,
        selectedTreeIds: action.mode === 'add' ? union(state.selectedTreeIds, action.treeIds) : action.treeIds,
        selectedNodeIds: action.mode === 'add' ? union(state.selectedNodeIds, action.nodeIds) : action.nodeIds,
      };
    default:
      return state;
  }
};
