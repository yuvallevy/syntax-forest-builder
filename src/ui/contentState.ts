import {
  Id, IdMap, StringSlice, Sentence, UnpositionedPlot, UnpositionedTree, TreeAndNodeId, PlotCoordsOffset
} from '../core/types';
import { deleteNodesInTree, InsertedNode, insertNodeIntoTree, transformNodeInTree } from '../mantle/manipulation';
import UndoRedoHistory, { ApplyActionFunc, applyToHistory, redo, ReverseActionFunc, undo, UndoableActionCommon } from '../mantle/UndoRedoHistory';
import { handleLocalSentenceChange } from './editNodes';
import { omitKey } from '../core/objTransforms';

/**
 * Represents an action taken by the user.
 * To integrate with the undo/redo system, each action is translated into an undoable "state change",
 * which is subsequently applied in a reversible fashion.
 */
export type ContentAction =
  | { type: 'insertNode', plotId: Id, treeId: Id, newNodeId: Id, newNode: InsertedNode }
  | { type: 'deleteNodes', plotId: Id, nodes: TreeAndNodeId[] }
  | { type: 'setNodeLabel', plotId: Id, node: TreeAndNodeId, newLabel: string }
  | { type: 'setSentence', plotId: Id, treeId: Id, newSentence: Sentence, oldSelectedSlice: StringSlice }
  | { type: 'addTree', plotId: Id, newTreeId: Id, offset: PlotCoordsOffset }
  | { type: 'removeTree', plotId: Id, treeId: Id }
;

/**
 * Represents a change in state as a result of an action done by the user.
 * Each state change includes information about the state before it so it can be easily undone,
 * and each action by the user is translated into a state change so that undo/redo can work smoothly.
 */
type ContentChange = (
  | { type: 'setTree', plotId: Id, treeId: Id, old: UnpositionedTree, new: UnpositionedTree }
  | { type: 'addTree', plotId: Id, newTreeId: Id, newTree: UnpositionedTree }
  | { type: 'removeTree', plotId: Id, treeId: Id, removedTree: UnpositionedTree }
);

type UndoableContentChange = UndoableActionCommon & ContentChange;

/**
 * Translates a user action into a state change, which can later be undone.
 */
const makeUndoable = (state: ContentState) => (action: ContentAction): ContentChange => {
  switch (action.type) {
    case 'insertNode':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: insertNodeIntoTree(action.newNode)(action.newNodeId)(state.plots[action.plotId].trees[action.treeId]),
      };
    case 'deleteNodes':
      const treeId = action.nodes[0].treeId;  // TODO: Use all trees
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: treeId,
        old: state.plots[action.plotId].trees[treeId],
        new: deleteNodesInTree(action.nodes.map(({ nodeId }) => nodeId))(state.plots[action.plotId].trees[treeId]),
      };
    case 'setNodeLabel':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.node.treeId,
        old: state.plots[action.plotId].trees[action.node.treeId],
        new: transformNodeInTree(node => ({ ...node, label: action.newLabel }))(action.node.nodeId)(
          state.plots[action.plotId].trees[action.node.treeId]),
      };
    case 'setSentence':
      return {
        type: 'setTree',
        plotId: action.plotId,
        treeId: action.treeId,
        old: state.plots[action.plotId].trees[action.treeId],
        new: handleLocalSentenceChange(action.newSentence, action.oldSelectedSlice)(
          state.plots[action.plotId].trees[action.treeId]),
      };
    case 'addTree':
      return {
        type: 'addTree',
        plotId: action.plotId,
        newTreeId: action.newTreeId,
        newTree: {
          nodes: {},
          sentence: '',
          offset: action.offset,
        },
      };
    case 'removeTree':
      return {
        ...action,
        removedTree: state.plots[action.plotId].trees[action.treeId],
      };
  }
};

const applyUndoableAction: ApplyActionFunc<UndoableContentChange, ContentState> = action => state => {
  switch (action.type) {
    case 'setTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: {
              ...state.plots[action.plotId].trees,
              [action.treeId]: action.new,
            },
          },
        },
      };
    case 'addTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: {
              ...state.plots[action.plotId].trees,
              [action.newTreeId]: action.newTree,
            },
          },
        },
      };
    case 'removeTree':
      return {
        ...state,
        plots: {
          ...state.plots,
          [action.plotId]: {
            ...state.plots[action.plotId],
            trees: omitKey(state.plots[action.plotId].trees, action.treeId)
          },
        },
      };
    default:
      return state;
  }
};

const reverseUndoableAction: ReverseActionFunc<UndoableContentChange> = action => {
  switch (action.type) {
    case 'setTree':
      return {
        ...action,
        old: action.new,
        new: action.old,
      };
    case 'addTree':
      return {
        ...action,
        type: 'removeTree',
        plotId: action.plotId,
        treeId: action.newTreeId,
        removedTree: action.newTree,
      };
    case 'removeTree':
      return {
        ...action,
        type: 'addTree',
        plotId: action.plotId,
        newTreeId: action.treeId,
        newTree: action.removedTree,
      };
    default:
      return action;
  }
};

export type ContentState = {
  plots: IdMap<UnpositionedPlot>;
};

export type UndoableContentState = UndoRedoHistory<UndoableContentChange, ContentState>;

const initialState: ContentState = {
  plots: {
    'plot': {
      trees: {},
    },
  },
};

export const initialContentState: UndoableContentState = {
  current: initialState,
  undoStack: [],
  redoStack: [],
};

export const contentReducer = (state: UndoableContentState, action: ContentAction | { type: 'undo' } | { type: 'redo' }): UndoableContentState =>
  action.type === 'undo' ? undo(applyUndoableAction)(reverseUndoableAction)(state)
    : action.type === 'redo' ? redo(applyUndoableAction)(reverseUndoableAction)(state)
    : applyToHistory(applyUndoableAction)({ ...makeUndoable(state.current)(action), timestamp: new Date() })(state);
