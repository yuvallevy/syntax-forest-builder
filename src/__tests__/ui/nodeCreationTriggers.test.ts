import { describe, expect, it } from 'vitest';
import { getNodeCreationTriggersForTree } from '../../ui/nodeCreationTriggers';
import mockStrWidth from '../__mocks__/mockStrWidth';
import { PositionedTree } from '../../content/positioned/types';

describe('node creation triggers', () => {
  const treeWithNoUnassignedSlices: PositionedTree = {
    sentence: 'Noun verbs very adverbly.',
    nodes: {
      'b': { label: 'NP', position: { treeX: 18, treeY: -60 }, children: ['c'] },
      'c': { label: 'N', position: { treeX: 18, treeY: -2 }, slice: [0, 4] },
      'd': { label: 'VP', position: { treeX: 89.25, treeY: -60 }, children: ['e', 'f'] },
      'e': { label: 'V', position: { treeX: 57, treeY: -2 }, slice: [5, 10] },
      'f': { label: 'AdvP', position: { treeX: 121.5, treeY: -30 }, triangle: { treeX1: 72, treeX2: 104 }, slice: [11, 24] },
    },
    position: { plotX: 50, plotY: -32 },
    width: 104,
  };
  const treeWithUnassignedSlices: PositionedTree = {
    sentence: 'Noun verbs very adverbly.',
    nodes: {
      'b': { label: 'NP', position: { treeX: 18, treeY: -60 }, children: ['c'] },
      'c': { label: 'N', position: { treeX: 18, treeY: -2 }, slice: [0, 4] },
      'd': { label: 'VP', position: { treeX: 57, treeY: -60 }, children: ['e'] },
      'e': { label: 'V', position: { treeX: 57, treeY: -2 }, slice: [5, 10] },
    },
    position: { plotX: 50, plotY: -32 },
    width: 104,
  };

  it('generates node creation triggers for a tree with all slices assigned', () => {
    expect(getNodeCreationTriggersForTree(mockStrWidth)(treeWithNoUnassignedSlices)).toMatchSnapshot();
  });

  it('generates node creation triggers for a tree with unassigned slices', () => {
    expect(getNodeCreationTriggersForTree(mockStrWidth)(treeWithUnassignedSlices)).toMatchSnapshot();
  });
});
