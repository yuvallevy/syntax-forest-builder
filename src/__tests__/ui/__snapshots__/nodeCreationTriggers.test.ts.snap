// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`node creation triggers > generates node creation triggers for a tree with all slices assigned 1`] = `
[
  {
    "bottomRight": {
      "treeX": 34,
      "treeY": -80,
    },
    "childIds": [
      "b",
    ],
    "childPositions": [
      {
        "treeX": 18,
        "treeY": -60,
      },
    ],
    "origin": {
      "treeX": 18,
      "treeY": -100,
    },
    "topLeft": {
      "treeX": 2,
      "treeY": -128,
    },
  },
  {
    "bottomRight": {
      "treeX": 105.25,
      "treeY": -80,
    },
    "childIds": [
      "d",
    ],
    "childPositions": [
      {
        "treeX": 89.25,
        "treeY": -60,
      },
    ],
    "origin": {
      "treeX": 89.25,
      "treeY": -100,
    },
    "topLeft": {
      "treeX": 73.25,
      "treeY": -128,
    },
  },
  {
    "bottomRight": {
      "treeX": 69.625,
      "treeY": -80,
    },
    "childIds": [
      "b",
      "d",
    ],
    "childPositions": [
      {
        "treeX": 18,
        "treeY": -60,
      },
      {
        "treeX": 89.25,
        "treeY": -60,
      },
    ],
    "origin": {
      "treeX": 53.625,
      "treeY": -100,
    },
    "topLeft": {
      "treeX": 37.625,
      "treeY": -128,
    },
  },
]
`;

exports[`node creation triggers > generates node creation triggers for a tree with unassigned slices 1`] = `
[
  {
    "bottomRight": {
      "treeX": 34,
      "treeY": -80,
    },
    "childIds": [
      "b",
    ],
    "childPositions": [
      {
        "treeX": 18,
        "treeY": -60,
      },
    ],
    "origin": {
      "treeX": 18,
      "treeY": -100,
    },
    "topLeft": {
      "treeX": 2,
      "treeY": -128,
    },
  },
  {
    "bottomRight": {
      "treeX": 73,
      "treeY": -80,
    },
    "childIds": [
      "d",
    ],
    "childPositions": [
      {
        "treeX": 57,
        "treeY": -60,
      },
    ],
    "origin": {
      "treeX": 57,
      "treeY": -100,
    },
    "topLeft": {
      "treeX": 41,
      "treeY": -128,
    },
  },
  {
    "bottomRight": {
      "treeX": 53.5,
      "treeY": -80,
    },
    "childIds": [
      "b",
      "d",
    ],
    "childPositions": [
      {
        "treeX": 18,
        "treeY": -60,
      },
      {
        "treeX": 57,
        "treeY": -60,
      },
    ],
    "origin": {
      "treeX": 37.5,
      "treeY": -100,
    },
    "topLeft": {
      "treeX": 21.5,
      "treeY": -128,
    },
  },
  {
    "bottomRight": {
      "treeX": 108,
      "treeY": 0,
    },
    "origin": {
      "treeX": 92,
      "treeY": -20,
    },
    "slice": [
      11,
      15,
    ],
    "topLeft": {
      "treeX": 76,
      "treeY": -48,
    },
  },
  {
    "bottomRight": {
      "treeX": 153.5,
      "treeY": 0,
    },
    "origin": {
      "treeX": 137.5,
      "treeY": -20,
    },
    "slice": [
      16,
      24,
    ],
    "topLeft": {
      "treeX": 121.5,
      "treeY": -48,
    },
  },
]
`;
